File: 1, Arrays and Strings/8/ZeroMatrix.java
// Write an algorithm such that if an element in an MxN matrix is 0,
// its entire row and column are set to 0

class ZeroMatrix{
    public static void zeroMatrix(int[][] matrix){
        Integer[][] copy = new Integer[matrix.length][matrix[0].length];
        for(int y=0; y<matrix.length; y++){
            for(int x=0; x<matrix[y].length; x++){
                if(copy[y][x] == null){
                    copy[y][x] = matrix[y][x];
                }
                if(matrix[y][x] == 0){
                    System.out.println("found 0");
                    for(int xi=0; xi<matrix[y].length; xi++){
                        copy[y][xi] = 0;
                    }
                    for(int yi=0; yi<matrix.length; yi++){
                        copy[yi][x] = 0;
                    }
                    print(copy);
                }
            }
        }
        for(int y=0; y<matrix.length; y++){
            for(int x=0; x<matrix[y].length; x++){
                matrix[y][x] = copy[y][x];
            }
        }
    }

    public static void main(String[] args){
        int[][] matrix = {
            {1,2,3,4},
            {5,6,7,8},
            {9,0,1,2}
        };
        print(matrix);
        zeroMatrix(matrix);
        print(matrix);
    }

    public static void print(int[][] matrix){
        for(int y=0; y<matrix.length; y++){
            for(int x=0; x<matrix[y].length; x++){
                System.out.print(matrix[y][x]+", ");
            }
            System.out.println();
        }
        System.out.println();
    }
    
    public static void print(Integer[][] matrix){
        for(int y=0; y<matrix.length; y++){
            for(int x=0; x<matrix[y].length; x++){
                System.out.print(matrix[y][x]+", ");
            }
            System.out.println();
        }
        System.out.println();
    }
}

File: 1, Arrays and Strings/8/ZeroMatrix.js
// Zero Matrix
class Matrix {
  constructor(m, n) {
    this.matrix = [];
    for (let i = 0; i < m; i++) {
      this.matrix[i] = [];
    }
  }

  zeroMatrix() {
    const zeroes = this.getZeroes();
    zeroes.forEach(zero => {
      for (let i = 0; i < this.matrix[0].length; i++) {
        this.matrix[zero.y][i] = 0;
      }
      for (let i = 0; i < this.matrix.length; i++) {
        this.matrix[i][zero.x] = 0;
      }
    });
  }

  getZeroes() {
    const zeroes = [];
    for (let i = 0; i < this.matrix.length; i++) {
      for (let j = 0; j < this.matrix[i].length; j++) {
        if (this.matrix[i][j] == 0) zeroes.push({ x: j, y: i });
      }
    }
    return zeroes;
  }

  printMatrix() {
    this.matrix.forEach(row => {
      console.log(row);
    });
  }
}

const matrix = new Matrix(3, 3);
matrix.matrix[0] = [1, 2, 3, 4];
matrix.matrix[1] = [1, 2, 3, 4];
matrix.matrix[2] = [0, 2, 3, 4];
console.log(matrix.getZeroes());
matrix.printMatrix();
matrix.zeroMatrix();
console.log('Matrix zeroed');
matrix.printMatrix();
